src/platforms/pipedrive/client.ts

```diff
@@ -0,0 +1,193 @@
+import { Auth, ClientResult } from '@/sdk';
+import {
+  formatUpsertInputWithNative,
+  formatUrl,
+  makeRequestFactory,
+} from '@/sdk/client';
+import { omit, shake } from 'radash';
+import * as z from 'zod';
+import { API_VERSION, BASE_URL, PIPEDRIVE_MAX_PAGE_SIZE } from './constants';
+import {
+  BatchReadObjectInput,
+  FindObjectInput,
+  ListObjectInput,
+  ListOutput,
+  listResponseSchema,
+  PipedriveDeal,
+  PipedriveDealCreate,
+  pipedriveDealSchema,
+  PipedriveDealUpdate,
+  PipedriveModule,
+  PipedriveNote,
+  PipedriveNoteCreate,
+  pipedriveNoteSchema,
+  PipedriveNoteUpdate,
+  PipedrivePerson,
+  PipedrivePersonCreate,
+  PipedrivePersonUpdate,
+  PipedriveUser,
+  pipedriveUserSchema,
+  SearchObjectInput,
+  SearchOutput,
+  searchResponseSchema,
+  UpsertResponse,
+  upsertResponseSchema,
+} from './schemas';
+
+const request = makeRequestFactory(async (auth, options) => {
+  return {
+    ...options,
+    url: formatUrl(BASE_URL, options.url),
+    headers: {
+      ...options.headers,
+      Authorization: `Bearer ${await auth.getToken()}`,
+    },
+  };
+});
+
+type requestFunctionType<I, O> = (
+  auth: Auth,
+  input: I,
+) => Promise<ClientResult<O>>;
+
+const makeClient = () => {
+  const findObject = <TOutput>(
+    module: PipedriveModule,
+    schema: z.ZodSchema,
+    properties?: string[],
+  ): requestFunctionType<FindObjectInput, TOutput> =>
+    request(({ id }: FindObjectInput) => ({
+      url: `/${API_VERSION}/${module}/${id}`,
+      method: 'GET',
+      schema,
+    }));
+
+  const listObject = <TOutput>(
+    module: PipedriveModule,
+    schema: z.ZodSchema,
+  ): requestFunctionType<ListObjectInput, ListOutput<TOutput>> =>
+    request(({ start, limit = PIPEDRIVE_MAX_PAGE_SIZE }: ListObjectInput) => ({
+      url: `/${API_VERSION}/${module}`,
+      method: 'GET',
+      query: shake({
+        start,
+        limit,
+      }),
+      schema: listResponseSchema(schema),
+    }));
+
+  const createObject = <TInput extends Record<string, unknown>>(
+    module: PipedriveModule,
+  ): requestFunctionType<TInput, UpsertResponse> =>
+    request((body: TInput) => ({
+      url: `/${API_VERSION}/${module}/`,
+      method: 'POST',
+      schema: upsertResponseSchema,
+      json: formatUpsertInputWithNative(body),
+    }));
+
+  const updateObject = <TInput extends Record<string, unknown>>(
+    module: PipedriveModule,
+  ): requestFunctionType<TInput, UpsertResponse> =>
+    request((body: TInput) => ({
+      url: `/${API_VERSION}/${module}/${body.id}`,
+      method: 'PUT',
+      schema: upsertResponseSchema,
+      json: formatUpsertInputWithNative(omit(body, ['id'])),
+    }));
+
+  const deleteObject = (
+    module: PipedriveModule,
+  ): requestFunctionType<FindObjectInput, void> =>
+    request((body: FindObjectInput) => ({
+      url: `/${API_VERSION}/${module}/${body.id}`,
+      method: 'DELETE',
+      schema: z.undefined(),
+    }));
+
+  const batchReadObject = <TOutput>(
```

Comment:
interesting, I'm thinking it's better to keep the client interface as close as possible to the interface of the downstream integration. Any particular reason to invent the new endpoint?

Reply:
Ah I figured since we're considering having the client be a standalone thing, having a batch functionality might make sense, since it's not obvious how to use the search query. But I agree, if it were easy to use the search query, I would remove this.

The SF client is kind of a good example of how we make things kind of easier based around common requests